pandas/core/indexes/base.py

```diff
@@ -6515,10 +6522,17 @@ def insert(self, loc: int, item) -> Index:
             loc = loc if loc >= 0 else loc - 1
             new_values[loc] = item
 
-        if self._typ == "numericindex":
-            # Use self._constructor instead of Index to retain NumericIndex GH#43921
-            # TODO(2.0) can use Index instead of self._constructor
-            return self._constructor._with_infer(new_values, name=self.name)
+        if not self._is_backward_compat_public_numeric_index:
+            from pandas.core.indexes.numeric import NumericIndex
+
+            if not isinstance(self, ABCRangeIndex) or not isinstance(
+                self, NumericIndex
+            ):
```

Comment:
doesnt `not isinstance(self, NumericIndex)` imply `not isinstance(self, ABCRangeIndex)`?

Reply:
Yeah, that's true. I'll fix it in the first follow-up commit.